_source/_posts/2024-04-30-spring-api-fga.md

```diff
@@ -0,0 +1,901 @@
+
+
+```shell
+fga model transform --file=auth-model.fga > auth-model.json
+```
+
+The transformed model should be like the JSON below:
+
+```json
+// auth-model.json
+{
+   "schema_version":"1.1",
+   "type_definitions":[
+      {
+         "type":"user"
+      },
+      {
+         "metadata":{...},
+         "relations":{...},
+         "type":"document"
+      },
+      {
+         "metadata":{...},
+         "relations":{...},
+         "type":"domain"
+      }
+   ]
+}
+```
+
+Place the file at `src/main/resources/fga/auth-model.json` as it will be required later.
+
+## Add fine-grained authorization (FGA) with OpenFGA
+
+Now let's integrate fine-grained authorization into the application. The OpenFGA team has just released the OpenFGA Spring Boot Starter 0.0.1, and you can add it to your project with the following dependency:
+
+```groovy
+implementation 'dev.openfga:openfga-spring-boot-starter:0.0.1'
+```
+
+The OpenFGA Spring Boot Starter dependency provides the auto-configuration of an FGA client and FGA bean that exposes a check method for authorizing operations. Before adding method-security, let's add the changes required for creating the owner tuple when the document is created. Create an `AuthorizationService` in the package `com.example.demo.service`:
+
+```java
+// src/main/java/com/example/demo/service/AuthorizationService.java
+package com.example.demo.service;
+
+import com.example.demo.model.Permission;
+import dev.openfga.sdk.api.client.OpenFgaClient;
+import dev.openfga.sdk.api.client.model.ClientTupleKey;
+import dev.openfga.sdk.api.client.model.ClientWriteResponse;
+import dev.openfga.sdk.errors.FgaInvalidParameterException;
+import org.springframework.stereotype.Service;
+
+import java.util.List;
+import java.util.concurrent.ExecutionException;
+
+@Service
+public class AuthorizationService {
+
+    private OpenFgaClient fgaClient;
+
+    public AuthorizationService(OpenFgaClient fgaClient) {
+        this.fgaClient = fgaClient;
+    }
+
+    public void create(Permission permission) {
+        try {
+            ClientTupleKey tuple = new ClientTupleKey()
+                    .user("user:" + permission.getUserId())
+                    .relation(permission.getRelation())
+                    ._object("document:" + permission.getDocumentId());
+            ClientWriteResponse response = fgaClient.writeTuples(List.of(tuple)).get();
+        } catch (FgaInvalidParameterException | InterruptedException | ExecutionException e) {
+            throw new AuthorizationServiceException(e);
```

Comment:
what is the purpose of wrapping the exceptions here, the AuthorizationServiceException doesnt seem to do anything

Reply:
For using the exception type in [PermissionController](https://github.com/indiepopart/spring-api-fga/blob/main/final/src/main/java/com/example/demo/web/PermissionController.java) @ExceptionHandler. How can it be improved?

Reply:
ah ok, I overlooked that, its fine then